scc/light_client/retry.go

```diff
@@ -0,0 +1,121 @@
+// number of retries and total timeout.
+//
+// Parameters:
+//   - provider: The underlying provider to wrap with retry logic.
+//   - retries: The maximum number of attempt to ask for certificates.
+//   - timeout: The time max time willing to wait. if it is zero value,
+//     then timeout is 10 seconds.
+//
+// Returns:
+// - *retryProvider: A new retryProvider provider instance.
+func newRetry(provider provider, retries uint, timeout time.Duration) *retryProvider {
+	if timeout == 0 {
+		timeout = 10 * time.Second
+	}
+	return &retryProvider{
+		provider: provider,
+		retries:  retries,
+		timeout:  timeout,
+	}
+}
+
+// Close closes the retryProvider.
+// Closing an already closed provider has no effect.
+func (r *retryProvider) close() {
+	r.provider.close()
+}
+
+// getCommitteeCertificates returns up to `maxResults` consecutive committee
+// certificates starting from the given period.
+//
+// Parameters:
+// - first: The starting period for which to retrieve committee certificates.
+// - maxResults: The maximum number of committee certificates to retrieve.
+//
+// Returns:
+// - []cert.CommitteeCertificate: A slice of committee certificates.
+// - error: An error if the provider failed to obtain the requested certificates.
+func (r retryProvider) getCommitteeCertificates(first scc.Period, maxResults uint64) ([]cert.CommitteeCertificate, error) {
+	return retry(r.retries, r.timeout, func() ([]cert.CommitteeCertificate, error) {
+		return r.provider.getCommitteeCertificates(first, maxResults)
+	})
+}
+
+// getBlockCertificates returns up to `maxResults` consecutive block
+// certificates starting from the given block number.
+//
+// Parameters:
+// - first: The starting block number for which to retrieve the block certificate.
+// - maxResults: The maximum number of block certificates to retrieve.
+//
+// Returns:
+//   - []cert.BlockCertificate: The block certificates for the given block number
+//     and the following blocks.
+//   - error: An error if the provider failed to obtain the requested certificates.
+func (r retryProvider) getBlockCertificates(first idx.Block, maxResults uint64) ([]cert.BlockCertificate, error) {
+	return retry(r.retries, r.timeout, func() ([]cert.BlockCertificate, error) {
+		return r.provider.getBlockCertificates(first, maxResults)
+	})
+}
+
+// retry executes the given function a number of times equal to retries+1, unless
+// one it returns a nil error, with incremental delays, waiting up to a max of timeout.
+//
+// Parameters:
+// - fn: The function to execute and retry if failed.
+//
+// Returns:
+//   - C: The result of the function if it succeeded.
+//   - error: Nil if at least one execution of fn returned without error.
+//     The joined error of all failed retries if all calls to fn failed.
+func retry[C any](retries uint, timeout time.Duration, fn func() (C, error)) (C, error) {
+	var errs []error
+	now := time.Now()
+	delay := time.Millisecond
+	maxDelay := 128 * time.Millisecond
```

Comment:
nit: this timing is a bit aggressive when addressing a real web-server, as it would produce ~8 queries per second at the minimum. I would increase it to more like a second or two for the `maxDelay`. Also the initial delay should be increased to ~200 milliseconds, as otherwise it would just send the initial few queries in very quick succession.

Note: the maximum delay and latency in the unit test example I was providing a link to is intended to be very short as it only sends those requests to the local process; these requests do not even leave the address space, and thus can be set up at a higher rate to minimize delay during unit testing.

Reply:
initial delayed changed to 200 miliseconds. max delay capped at 1 second. https://github.com/0xsoniclabs/sonic/pull/128/commits/1be35a75a4752a671ca398509b4002d0b30fac7f